In [3]:
pip install keras


Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install tensorflow


     |████████████████████████████████| 165.1 MB 69 kB/s  eta 0:00:01    |██████████████████▎             | 94.6 MB 4.2 MB/s eta 0:00:17     |███████████████████████████████▋| 163.2 MB 4.1 MB/s eta 0:00:01
     |████████████████████████████████| 6.8 MB 4.2 MB/s eta 0:00:01
     |████████████████████████████████| 3.4 MB 4.6 MB/s eta 0:00:01
     |████████████████████████████████| 42 kB 2.6 MB/s  eta 0:00:01
     |████████████████████████████████| 65 kB 5.1 MB/s eta 0:00:011
     |████████████████████████████████| 127 kB 5.7 MB/s eta 0:00:01
     |████████████████████████████████| 1.3 MB 5.5 MB/s eta 0:00:01
     |████████████████████████████████| 459 kB 3.8 MB/s eta 0:00:01
     |████████████████████████████████| 57 kB 6.2 MB/s eta 0:00:011
     |████████████████████████████████| 95 kB 5.2 MB/s eta 0:00:011
     |████████████████████████████████| 114 kB 5.1 MB/s eta 0:00:01
     |████████████████████████████████| 779 kB 5.1 MB/s eta 0:00:01
     |████████████████████████████████| 155 kB

In [3]:
# first neural network with keras tutorial
import numpy as np
from keras.models import Sequential
from keras.layers import Dense
import pandas as pd

In [4]:
df=pd.read_csv('/Users/lenkwok/pCloud Drive/pima-indians-diabetes.data.csv')

In [56]:
df.head(10)

6  148  72  35    0  33.6  0.627  50  1
0   1   85  66  29    0  26.6  0.351  31  0
1   8  183  64   0    0  23.3  0.672  32  1
2   1   89  66  23   94  28.1  0.167  21  0
3   0  137  40  35  168  43.1  2.288  33  1
4   5  116  74   0    0  25.6  0.201  30  0
5   3   78  50  32   88  31.0  0.248  26  1
6  10  115   0   0    0  35.3  0.134  29  0
7   2  197  70  45  543  30.5  0.158  53  1
8   8  125  96   0    0   0.0  0.232  54  1
9   4  110  92   0    0  37.6  0.191  30  0

In [ ]:
#Input Variables (X):


## 1. Number of times pregnant
   2. Plasma glucose concentration a 2 hours in an oral glucose tolerance test
   3. Diastolic blood pressure (mm Hg)
   4. Triceps skin fold thickness (mm)
   5. 2-Hour serum insulin (mu U/ml)
   6. Body mass index (weight in kg/(height in m)^2)
   7. Diabetes pedigree function
   8. Age (years)
   9. Class variable (0 or 1)

#Output Variables (y):(0 or 1)

In [19]:
#Split dataset

X = df.iloc[:, :-1]
y=df.iloc[:,-1]


# Define Keras Model
#In this example, we will use a fully-connected network structure with three layers.

Fully connected layers are defined using the Dense class. We can specify the number of neurons or nodes in the layer as the first argument, and specify the activation function using the activation argument.

We will use the rectified linear unit activation function referred to as ReLU on the first two layers and the Sigmoid function in the output layer.

It used to be the case that Sigmoid and Tanh activation functions were preferred for all layers. 
These days, better performance is achieved using the ReLU activation function. We use a sigmoid on the output layer to ensure our network output is between 0 and 1 and easy to map to either a probability of class 1 or snap to a hard classification of either class with a default threshold of 0.5.

We can piece it all together by adding each layer:

The model expects rows of data with 8 variables (the input_dim=8 argument)
The first hidden layer has 12 nodes and uses the relu activation function.
The second hidden layer has 8 nodes and uses the relu activation function.
The output layer has one node and uses the sigmoid activation function.

In [21]:
# define the keras model
model = Sequential()
model.add(Dense(12, input_dim=8, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

# Compile Keras Model
Compiling the model uses the efficient numerical libraries under the covers (the so-called backend) 
such as Theano or TensorFlow. The backend automatically chooses the best way to represent the network for training and making predictions to run on your hardware, such as CPU or GPU or even distributed.

We must specify the loss function to use to evaluate a set of weights, the optimizer is used to search through different weights for the network and any optional metrics we would like to collect and report during training.

In this case, we will use cross entropy as the loss argument. This loss is for a binary classification problems and is defined in Keras as “binary_crossentropy“

We will define the optimizer as the efficient stochastic gradient descent algorithm “adam“. This is a popular version of gradient descent because it automatically tunes itself and gives good results in a wide range of problems. 

Finally, because it is a classification problem, we will collect and report the classification accuracy, defined via the metrics argument.

In [22]:
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])


# 4. Fit Keras Model
Training occurs over epochs and each epoch is split into batches.

Epoch: One pass through all of the rows in the training dataset.
Batch: One or more samples considered by the model within an epoch before weights are updated.

For this problem, we will run for a small number of epochs (150) and use a relatively small batch size of 10.

These configurations can be chosen experimentally by trial and error. We want to train the model enough so that it learns a good (or good enough) mapping of rows of input data to the output classification. The model will always have some error, but the amount of error will level out after some point for a given model configuration. This is called model convergence.


In [23]:
#fit the keras model on the dataset
model.fit(X, y, epochs=150, batch_size=10)

Epoch 1/150
77/77 [==============================] - 0s 566us/step - loss: 4.1207 - accuracy: 0.5098
Epoch 2/150
77/77 [==============================] - 0s 574us/step - loss: 1.4091 - accuracy: 0.5711
Epoch 3/150
77/77 [==============================] - 0s 557us/step - loss: 1.1283 - accuracy: 0.5984
Epoch 4/150
77/77 [==============================] - 0s 602us/step - loss: 0.9823 - accuracy: 0.6167
Epoch 5/150
77/77 [==============================] - 0s 566us/step - loss: 0.8317 - accuracy: 0.6310
Epoch 6/150
77/77 [==============================] - 0s 570us/step - loss: 0.7595 - accuracy: 0.6441
Epoch 7/150
77/77 [==============================] - 0s 601us/step - loss: 0.7079 - accuracy: 0.6597
Epoch 8/150
77/77 [==============================] - 0s 580us/step - loss: 0.6827 - accuracy: 0.6558
Epoch 9/150
77/77 [==============================] - 0s 559us/step - loss: 0.6669 - accuracy: 0.6780
Epoch 10/150
77/77 [==============================] - 0s 593us/step - loss: 0.6514 - accura

# 5. Evaluate Keras Model
The evaluate() function will return a list with two values. The first will be the loss of the model on the dataset and the second will be the accuracy of the model on the dataset. 


In [24]:
# evaluate the keras model
_, accuracy = model.evaluate(X, y)
print('Accuracy: %.2f' % (accuracy*100))


24/24 [==============================] - 0s 493us/step - loss: 0.4698 - accuracy: 0.7797
Accuracy: 77.97


# Make Predictions

In [25]:
# make class predictions with the model
predictions = model.predict_classes(X)

Instructions for updating:
Please use instead:* `np.argmax(model.predict(x), axis=-1)`,   if your model does multi-class classification   (e.g. if it uses a `softmax` last-layer activation).* `(model.predict(x) > 0.5).astype("int32")`,   if your model does binary classification   (e.g. if it uses a `sigmoid` last-layer activation).


The complete example below makes predictions for each example in the dataset, then prints the input data, predicted class and expected class for the first 5 examples in the dataset.

In [57]:
for i in range(10):
	print('predicted output %d,actual %d' % (predictions[i], y[i]))
  

predicted output 0,actual 0
predicted output 1,actual 1
predicted output 0,actual 0
predicted output 1,actual 1
predicted output 0,actual 0
predicted output 0,actual 1
predicted output 0,actual 0
predicted output 1,actual 1
predicted output 0,actual 1
predicted output 0,actual 0


The second last prediction is inaccurate, but we can see that most rows are correctly predicted. In fact, we would expect about 77.97% of the rows to be correctly predicted based on our estimated performance of the model.